pkg/controller/statefulset/stateful_set_control_test.go

```diff
@@ -607,6 +611,345 @@ func TestStatefulSetControl_getSetRevisions(t *testing.T) {
+	}
+	originalPods, err := spc.podsLister.Pods(set.Namespace).List(selector)
+	if err != nil {
+		t.Fatal(err)
+	}
+	sort.Sort(ascendingOrdinal(originalPods))
+
+	// first update pods 4/5, this will delete 4/5, since maxUnavailable is 2
+	if err = ssc.UpdateStatefulSet(set, originalPods); err != nil {
+		t.Fatal(err)
+	}
+	pods, err := spc.podsLister.Pods(set.Namespace).List(selector)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	sort.Sort(ascendingOrdinal(pods))
+
+	// expected number of pod is 0,1,2,3
+	if !reflect.DeepEqual(pods, originalPods[:4]) {
+		t.Fatalf("Expected pods %v, got pods %v", originalPods[:4], pods)
+	}
+
+	// create new pods 5(only one pod gets created at a time due to OrderedReady)
+	if err = ssc.UpdateStatefulSet(set, pods); err != nil {
+		t.Fatal(err)
+	}
+	pods, err = spc.podsLister.Pods(set.Namespace).List(selector)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(pods) != 5 {
+		t.Fatalf("Expected create pods 5, got pods %v", len(pods))
+	}
+	spc.setPodRunning(set, 4)
+	pods, _ = spc.setPodReady(set, 4)
+
+	// create new pods 4(only one pod gets created at a time due to OrderedReady)
+	if err = ssc.UpdateStatefulSet(set, pods); err != nil {
+		t.Fatal(err)
+	}
+	pods, err = spc.podsLister.Pods(set.Namespace).List(selector)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(pods) != 6 {
+		t.Fatalf("Expected create pods 4, got pods %v", len(pods))
+	}
+
+	// if pod 4 ready, start to update pod 3
+	spc.setPodRunning(set, 5)
+	originalPods, _ = spc.setPodReady(set, 5)
+	sort.Sort(ascendingOrdinal(originalPods))
+	if err = ssc.UpdateStatefulSet(set, originalPods); err != nil {
+		t.Fatal(err)
+	}
+	pods, err = spc.podsLister.Pods(set.Namespace).List(selector)
+	if err != nil {
+		t.Fatal(err)
+	}
+	sort.Sort(ascendingOrdinal(pods))
+
+	// verify the remaining pods are 0,1,2,4,5 (3 got deleted)
+	if !reflect.DeepEqual(pods, append(originalPods[:3], originalPods[4:]...)) {
+		t.Fatalf("Expected pods %v, got pods %v", append(originalPods[:3], originalPods[4:]...), pods)
+	}
+
+	// create new pod 3
+	if err = ssc.UpdateStatefulSet(set, pods); err != nil {
+		t.Fatal(err)
+	}
+	pods, err = spc.podsLister.Pods(set.Namespace).List(selector)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(pods) != 6 {
+		t.Fatalf("Expected create pods 2/3, got pods %v", pods)
+	}
+
+	// pods 3/4/5 ready, should not update other pods
+	spc.setPodRunning(set, 3)
+	spc.setPodRunning(set, 5)
+	spc.setPodReady(set, 5)
+	originalPods, _ = spc.setPodReady(set, 3)
+	sort.Sort(ascendingOrdinal(originalPods))
+	if err = ssc.UpdateStatefulSet(set, originalPods); err != nil {
+		t.Fatal(err)
+	}
+	pods, err = spc.podsLister.Pods(set.Namespace).List(selector)
+	if err != nil {
+		t.Fatal(err)
+	}
+	sort.Sort(ascendingOrdinal(pods))
+	if !reflect.DeepEqual(pods, originalPods) {
+		t.Fatalf("Expected pods %v, got pods %v", originalPods, pods)
+	}
+}
+
+func TestStatefulSetControlRollingUpdateWithMaxUnavailableInOrderedModeVerifyInvariant(t *testing.T) {
```

Comment:
@lavalamp  this is in the new test i added

Reply:
This can be combined with above test.

Reply:
agree, this will also be optimized in a later PR

Reply:
Ok, please add a TODO, the updateStatefulSet method is reasonably easy to mock for the purpose of unit test.